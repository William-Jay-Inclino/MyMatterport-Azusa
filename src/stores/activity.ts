import { useSpaceStore } from "./space";
import { defineStore } from "pinia";
import { ref, computed } from "vue";
import { ACTIVITY_TYPE, type Activity } from "@/stores/types";
import { faker } from "@faker-js/faker/locale/ja";
import dayjs from "dayjs";

export const useActivityStore = defineStore("activity", () => {
  // State
  const spaceStore = useSpaceStore();
  const activity = ref<Array<Activity>>([]);
  const activityInfo = ref({} as Activity);
  const now = dayjs();

  // Getters

  // gets events by a pre-selected Space
  // const selectedSpaceActvities = computed(() => {
  //   console.log(
  //     "List of Acitivities by Space UUID",
  //     spaceStore.selectedTreeItem.uuid
  //   );
  //   return activity.value.filter(
  //     (item) => item.uuid === spaceStore.selectedTreeItem.uuid
  //   );
  // });

  // filters activities depending on passed Space UUID
  const getActivitiesBySpaceId = computed(() => (payload: Activity) => {
    console.log("List of Acitivities by Space UUID", payload.space_uuid);
    const filteredActivities = activity.value.filter(
      (item) => item.space_uuid === payload.space_uuid
    );
    console.log(filteredActivities);
    return filteredActivities;
  });

  //for population of fake data only
  const getActivities = (count = 40) => {
    console.log("Activities");

    for (let i = 0; i < count; i++) {
      const actInfo = {} as Activity;

      if (i < randomNumber.value) {
        actInfo.space_uuid = spaceStore.currentNode?.id;
      } else {
        actInfo.space_uuid = faker.datatype.uuid();
      }
      actInfo.type = ACTIVITY_TYPE.TAG_CLICK;
      actInfo.created_on = now.format("YYYY/MM/DD");

      console.log("Activity Details", actInfo);
      // activity.value.push(actInfo)
      addActivity(actInfo);
    }
  };

  //for random total number of activities in a selected space
  //fake app usaage only
  const randomNumber = computed(() => {
    return faker.datatype.number({ min: 15, max: 24 });
  });

  // Methods
  const addActivity = (payload: Activity) => {
    console.log("Acitvity Saved", payload);
    activity.value.unshift(payload);
  };

  return {
    // state
    activity,

    // getters
    // selectedSpaceActvities,
    getActivitiesBySpaceId,

    // methods
    addActivity,
    getActivities, //for population of fake data only
    randomNumber, //for fake app usage only
  };
});
