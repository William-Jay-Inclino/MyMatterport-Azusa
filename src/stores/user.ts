import { ref, computed } from "vue";
import { defineStore } from "pinia";
import type { User, Role } from "@/stores/types";
import { colors } from "./config";
import { faker } from "@faker-js/faker";

export const useUserStore = defineStore("user", () => {
  faker.setLocale('de')

  // flags
  const mode = ref("edit");

  // state data
  const users = ref<Array<User>>([]);
  const selectedUser = ref<User | null>(null);
  const checkedUsers = ref([] as User);
  const userInfo = ref({} as User);
  const roles = ref<Array<Role>>([]);
  const isAddAdminConfirmed = ref(false);

  const showUserForm = ref(false);

  const getRoles = () => {
    const collection: Array<Role> = [];

    for (let k = 0; k < 8; k++) {
      const r = {} as Role;
      const rand = Math.floor(Math.random() * colors.length);

      r.uuid = faker.datatype.uuid();
      r.color = colors[rand];
      r.name = faker.name.jobArea();
      collection.push(r);
    }

    roles.value = collection;
  };

  const getUsers = (count = 20) => {
    console.log("=== STORE: getUsers ===");
    // faker or axios

    getRoles();

    for (let i = 0; i < count; i++) {
      const user = {} as User;
      //let roleList:Array<Role> = []

      const collection: Array<Role> = [];
      //Get random number of roles
      let nRoles = Math.floor(Math.random() * 3);

      user.uuid = faker.datatype.uuid();
      user.first_name = faker.name.firstName();
      user.last_name = faker.name.lastName();
      user.middle_name = faker.name.middleName();
      user.company_name = faker.company.companyName();
      user.email = faker.internet.email();
      user.mobile = faker.phone.number();
      user.city = faker.address.cityName();
      user.password = "password";
      for (let j = 0; j <= nRoles; j++) {
        collection.push(
          roles.value[Math.floor(Math.random() * roles.value.length)]
        );
      }

      //user.roles = roleList
      user.roles = collection;

      users.value.push(user);
    }
  };

  const addUser = () => {
    console.log("=== STORE: addUser ===", userInfo.value);
    userInfo.value.uuid = faker.datatype.uuid();
    users.value.unshift(userInfo.value);
    resetForm();
  };

  const updateUser = (payload: User) => {
    console.log("=== STORE: updateUser ===", userInfo.value);

    // get the user record to update
    const userNdx = users.value.findIndex(
      (item) => item.uuid === userInfo.value?.uuid
    );

    console.log("=== userNdx ===", userNdx);
    if (userNdx > -1) {
      users.value.splice(userNdx, 1, userInfo.value);
    }
    // resetForm()
  };

  const deleteUser = (payload: User) => {
    console.log(
      "=== STORE: deleteUser ===",
      JSON.stringify(selectedUser.value)
    );

    // find the user that we wanted to delete
    const userNdx = users.value.findIndex(
      (item) => item.uuid === selectedUser.value?.uuid
    );

    console.log("=== userNdx ===", userNdx);
    if (userNdx > -1) {
      users.value.splice(userNdx, 1);
    }
    resetForm();
  };

  const resetForm = () => {
    selectedUser.value = null;
    userInfo.value = {} as User;
    showUserForm.value = false;
  };

  const setSelectedUser = (payload: User) => {
    selectedUser.value = payload;
  };

  // ======================== COMPUTED ======================== //
  const getRandomUser = computed(() => {
    const index = faker.datatype.number({ min: 0, max: 19 });
    return users.value[index];
  });

  return {
    // flags
    mode,
    //state
    users,
    selectedUser,
    checkedUsers,
    userInfo,
    roles,
    isAddAdminConfirmed,

    // computed
    showUserForm,
    getRandomUser,

    // actions
    getUsers,
    addUser,
    updateUser,
    deleteUser,
    resetForm,

    //method
    setSelectedUser,

    getRoles,

    // getters
  };
});
