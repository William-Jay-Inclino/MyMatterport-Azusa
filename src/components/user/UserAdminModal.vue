<template>
  <div class="admin-modal">
    <form>
      <div class="modal-card" v-if="!userStore.isAddAdminConfirmed">
        <header class="modal-card-head title is-4">Add Admin User</header>
        <o-field horizontal class="mt-5">
          <span class="label-text">Mail Address</span
          ><o-input
            type="email"
            v-model="userStore.userForm.email"
            class="login-input"
            expanded
        /></o-field>

        <footer class="modal-card-foot mt-5">
          <div class="column is-4 is-offset-2">
            <button
              class="button is-rounded is-primary"
              @click.prevent="cancelAddAdmin(), $emit('close')"
            >
              <span>Cancel</span>
            </button>
          </div>
          <div class="column is-4 is-offset-1">
            <button
              class="button is-rounded is-dark"
              :disabled="!userStore.userForm.email"
              @click.prevent="confirmAdd()"
            >
              <span>Add Admin ( Available: {{ usersLength }}) </span>
            </button>
          </div>
        </footer>
      </div>
      <div class="modal-card" v-else>
        <header class="modal-card-head title is-4">Confirm</header>
        <o-field horizontal class="mt-5 has-text-centered prompt">
          <span class="label-text"
            >If added user as admin, the user can access all other
            user/section/facility/space data. recommend to add user on section/facility
            setting as possible. do you still want to add admin user?</span
          >
        </o-field>

        <footer class="modal-card-foot mt-5">
          <div class="column is-4 is-offset-2">
            <button class="button is-rounded is-primary">
              <span @click.prevent="cancelAddAdmin(), $emit('close')">Cancel</span>
            </button>
          </div>
          <div class="column is-4 is-offset-1">
            <button
              class="button is-rounded is-dark"
              @click.prevent="addAdmin(userStore.userForm), $emit('close')"
            >
              <span>Add Admin ( Available: {{ usersLength }})</span>
            </button>
          </div>
        </footer>
      </div>
    </form>
  </div>
</template>

<script setup lang="ts">
import { useUserStore } from "@/stores/user";
import { onMounted, ref, computed } from "vue";
import { User } from "@/stores/types.ts";

const userStore = useUserStore();

// Function
const confirmAdd = () => {
  userStore.isAddAdminConfirmed = true;
  console.log("Confirming Adding of Admin", userStore.isAddAdminConfirmed);
};

const addAdmin = (payload: User) => {
  console.log("Adding Admin", payload);
  userStore.addUser(payload);
  userStore.isAddAdminConfirmed = false;
};

const cancelAddAdmin = () => {
  userStore.showUserForm = false;
  console.log("Cancel Adding of Admin");
  userStore.userForm = {};
};

const usersLength = computed(() => {
  return userStore.users.length;
});

onMounted(() => {
  console.log("This is mounted");
});
</script>

<style scoped>
.prompt {
  width: 495px;
  height: 96px;
  margin-left: 4rem;
}
.modal-card {
  width: 100%;
  max-height: calc(100vh - 160px);
  overflow: auto;
  position: relative;
  background: white;
}

.modal-card-foot,
.modal-card-head {
  align-items: center;
  background-color: white;
  display: flex;
  flex-shrink: 0;
  justify-content: flex-start;
  padding: 20px;
  position: relative;
}
.modal-card-head {
  border-bottom: 1px solid white;
  border-top-left-radius: 6px;
  border-top-right-radius: 6px;
}
.modal-card-body {
  -webkit-overflow-scrolling: touch;
  background-color: #fff;
  flex-grow: 1;
  flex-shrink: 1;
  overflow: auto;
  padding: 25px;
  margin-bottom: 0rem !important;
  display: inherit;
}
.modal-card-foot {
  border-bottom-left-radius: 6px;
  border-bottom-right-radius: 6px;
  border-top: 1px solid #dbdbdb;
  border-top: none !important;
  border-radius: 8px !important;
}
.modal-card-title {
  color: #363636;
  flex-grow: 1;
  flex-shrink: 0;
  font-size: 18px;
  line-height: 1;
  margin: 0;
  font-weight: 800;
}
.modal-card-foot .o-button:not(:last-child) {
  margin-left: 0.5em;
}
</style>
