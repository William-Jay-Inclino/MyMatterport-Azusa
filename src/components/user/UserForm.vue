<template>
  <main>
    <h1 class="has-text-primary">User Detail</h1>

    <o-field label="Firstname" style="width: 300px">
      <o-input v-model="userStore.userForm.first_name" data-test="input-first-name">
      </o-input>
    </o-field>

    <o-field label="Lastname">
      <o-input v-model="userStore.userForm.last_name" data-test="input-last-name">
      </o-input>
    </o-field>

    <o-field label="Email">
      <o-input v-model="userStore.userForm.email" data-test="input-email"> </o-input>
    </o-field>

    <o-field label="Mobile">
      <o-input v-model="userStore.userForm.mobile" data-test="input-mobile"> </o-input>
    </o-field>

    <o-button @click="saveClick()" data-test="button-save">Save</o-button>
    <o-button @click="cancelClick()" data-test="button-cancel">Cancel</o-button>
    <o-button
      v-show="userStore.mode === 'edit'"
      @click="deleteClick()"
      data-test="button-delete"
      >Delete</o-button
    >
  </main>
</template>

<script setup lang="ts">
import { useUserStore } from "@/stores/user";
import { onMounted, onUnmounted, onBeforeUpdate } from "vue";

// reactive data

// const table = ref();
const userStore = useUserStore();

// event handlers
const saveClick = () => {
  console.log("== DISPATCH: saveClick ===", userStore.mode);
  if (userStore.mode == "add") {
    userStore.addUser();
  } else {
    userStore.updateUser();
  }
};

const deleteClick = () => {
  console.log("== DISPATCH: deleteClick ===");
  userStore.deleteUser();
};

const cancelClick = () => {
  console.log("== DISPATCH: cancelClick ===");
  userStore.resetForm();
};

// lifecycle hooks function
onMounted(() => {
  console.log("=== onMounted === ");
});

onBeforeUpdate(() => {
  console.log("=== beforeUpdate === ");
});

onUnmounted(() => {
  console.log("=== onUnmounted === ");
});
</script>
