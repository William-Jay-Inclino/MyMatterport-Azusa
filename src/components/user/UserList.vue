<template>
  <div class="container">
    <div class="m-6">
      <h3 class="title is-1">UserList</h3>
      <o-field class="is-grouped">
        <o-input placeholder="<Search Input>" class="search-input" />

        <span class="mx-3">by</span>
        <select class="search-input">
          <option></option>
        </select>
      </o-field>
      <o-field class="add-icon is-block-absolute" v-if="!userStore.showUserForm">
        <button
          class="button is-ghost"
          @click="emit('add-user', { action: addUser, User: (userStore.userForm = {}) })"
        >
          <o-icon pack="mdi" icon="plus-circle-outline" size="small" class="button-icon">
          </o-icon></button
      ></o-field>
      <div class="column m-6">
        <o-table
          :data="userStore.users"
          v-model:checked-rows="userStore.checkedUsers"
          v-model:selected="userStore.selectedUser"
          focusable
          checkable
          :checkbox-position="checkboxPosition"
          :paginated="true"
          :per-page="8"
          :pagination-simple="true"
        >
          <o-table-column field="Name" width="150" label="Name" sortable #default="props"
            >{{ props.row.first_name + " " + props.row.last_name }}
          </o-table-column>
          <o-table-column
            field="Email"
            width="250"
            label="Email"
            sortable
            #default="props"
            >{{ props.row.email }}
          </o-table-column>
          <o-table-column
            field="Company"
            width="250"
            label="Company"
            sortable
            #default="props"
            >{{ props.row.company_name }}
          </o-table-column>
        </o-table>
      </div>
      <div class="has-text-centered">
        <button
          class="button is-rounded add-user"
          @click="
            emit('delete-user', { action: deleteUser, user: userStore.checkedUsers })
          "
        >
          <span class="menu-text">選択ユーザーを削除する</span
          ><o-icon pack="mdi" icon="account-minus-outline" class="mr-0"></o-icon>
        </button>
      </div>
    </div>
  </div>
</template>

<script setup lang="ts">
// Stores
import { onMounted, ref } from "vue";
import { useUserStore } from "@/stores/user";
import { useSessionStore } from "@/stores/session";

// Components
import TheLeftSidebar from "@/components/TheLeftSidebar.vue";

const userStore = useUserStore();
const checkboxPosition = ref("right");
const emit = defineEmits("add-user, delete-user");

userStore.getUsers();
const sessionStore = useSessionStore();

sessionStore.init();

onMounted(() => {
  console.log("UserList is mounted");
});
</script>

<style scoped>
.search-input {
  width: 66px !important;
}
.is-grouped {
  position: absolute !important;
  right: 50px;
  top: 26px;
}
.menu-width {
  margin-top: 0px;
}
.menu-text {
  margin-right: 10px;
}
.add-user {
  top: 0 !important;
  right: 50px;
}
</style>
