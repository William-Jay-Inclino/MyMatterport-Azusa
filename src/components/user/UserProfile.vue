<template>
  <div class="container">
    <div class="column is-hidden-tablet">
      <button
        class="button is-rounded"
        @click="spaceStore.isMobileSidebarActive = true"
        style="width: 80px"
      >
        <o-icon pack="mdi" icon="forwardburger" size="medium" class="button-icon">
        </o-icon>
      </button>
    </div>
    <div
      class="column profile-container is-three-quarters-mobile is-mobile"
      v-if="sessionStore.authUser"
    >
      <div class="is-block-absolute">
        <h1 class="is-4 profile-title has-text-black">Edit Profile</h1>
      </div>
      <div
        class="column profile-area is-three-fifths is-offset-one-fifth is-three-quarters-mobile"
      >
        <div class="password-wrapper">
          <o-field horizontal>
            <div class="label-text has-text-dark">Family Name</div>
            <o-input
              placeholder="Admin"
              class="login-input"
              expanded
              v-model="sessionStore.authUser.last_name"
            ></o-input>
          </o-field>
        </div>
        <div class="password-wrapper">
          <o-field horizontal>
            <div class="label-text has-text-dark">Name</div>
            <o-input
              placeholder="Administrator"
              class="login-input"
              expanded
              v-model="sessionStore.authUser.first_name"
            ></o-input>
          </o-field>
        </div>
        <div class="password-wrapper">
          <o-field horizontal>
            <div class="label-text has-text-dark">Middle Name</div>
            <o-input
              placeholder="middle name"
              class="login-input"
              expanded
              v-model="sessionStore.authUser.middle_name"
            ></o-input>
          </o-field>
        </div>
        <div class="password-wrapper">
          <o-field horizontal>
            <div class="label-text has-text-dark">Company Name</div>
            <o-input
              placeholder="Company Name"
              class="login-input"
              expanded
              v-model="sessionStore.authUser.company_name"
            ></o-input>
          </o-field>
        </div>
        <div class="password-wrapper">
          <o-field horizontal>
            <div class="label-text has-text-dark">Mail Address</div>
            <o-input
              placeholder="Mail Address"
              class="login-input"
              expanded
              disabled
              style="background: none; border: none"
              v-model="sessionStore.authUser.email"
            ></o-input>
          </o-field>
        </div>
        <div class="">
          <div class="password-wrapper">
            <o-field horizontal>
              <div class="label-text has-text-dark">Password</div>
              <button
                class="button is-primary"
                @click="emit('send-password', { action: 'Send Mail for Reset Password' })"
              >
                Send Reset Password to the Mail
              </button>
            </o-field>
          </div>
          <div>
            <button
              class="button is-rounded is-primary log-button"
              @click="
                emit('profile-update', {
                  action: 'Update Profile',
                  user: sessionStore.authUser,
                })
              "
            >
              <span class="menu-text">Update</span>
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</template>

<script setup lang="ts">
// Stores
import { useUserStore } from "@/stores/user";
import { useSessionStore } from "@/stores/session";
import { onMounted } from "vue";

// Components
import TheLeftSidebar from "@/components/TheLeftSidebar.vue";

const emit = defineEmits(["send-password", "profile-update"]);
const userStore = useUserStore();
const sessionStore = useSessionStore();

sessionStore.init();
</script>

<style scoped>
@media screen and (max-width: 768px) {
  .profile-area {
    top: 75px !important;
  }
  .profile-container {
    height: 715px;
    width: 395px;
    margin-top: calc(-6%);
    margin-left: calc(30%);
  }
  .log-button {
    left: -25px !important;
  }
}
.password-wrapper {
  margin-top: 20px !important;
}
.label-text {
  text-align: left;
}
.button {
  display: flex;
  width: 270px;
  height: 33px;
  border-radius: 50px;
}
.button-icon {
  margin-top: -30px !important;
}
.log-button {
  margin-left: 0px !important;
  right: 0;
  left: 0;
}
</style>
