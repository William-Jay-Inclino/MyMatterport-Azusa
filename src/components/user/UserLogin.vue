<template>
  <div class="columns is-full">
    <div class="">
      <div class="login-wrapper">
        <div class="logo-container m-auto">
          <img
            width="299"
            height="60"
            class="login-logo"
            src="/public/logo_archi_2.png"
          />
        </div>
        <div class="column is-block-absolute">
          <div class="column mail-wrapper">
            <o-field horizontal>
              <div class="label-text has-text-dark">Mail Address</div>
              <o-input
                v-model="userStore.userInfo.email"
                type="email"
                placeholder="sample@archi-twin.com"
                class="login-input"
                expanded
              ></o-input>
            </o-field>
          </div>
          <div class="column password-wrapper">
            <o-field horizontal>
              <div class="label-text has-text-dark">Password</div>
              <o-input
                v-model="userStore.userInfo.password"
                type="password"
                placeholder="password"
                class="login-input"
                expanded
                password-reveal
              ></o-input>
            </o-field>
          </div>
          <div class="column has-text-centered is-block-absolute">
            <span class="has-text-weight-medium has-text-danger error-message">
              {{ sessionStore.authErrorMessage }}
            </span>
          </div>
          <div>
            <button
              class="button is-rounded is-primary log-button"
              @click="emit('login-user', { user: userStore.userInfo })"
            >
              <span class="menu-text">Login</span>
              <o-icon
                pack="mdi"
                icon="login"
                size="small"
                class="button-icon has-text-white"
              >
              </o-icon>
            </button>
          </div>
        </div>
        <div class="column help-text-wrapper">
          <span class="help-text"
            >If you have forgotten your account information, please click here
            言語を変更したい場合は、ここをクリックしてください (JP) Powered by
            Architwin</span
          >
        </div>
      </div>
    </div>
  </div>
</template>

<script setup lang="ts">
// Stores
import UserForm from "@/components/user/UserForm.vue";
import { useSessionStore } from "@/stores/session";
import { useUserStore } from "@/stores/user";

const sessionStore = useSessionStore();
const userStore = useUserStore();

const emit = defineEmits(["login-user"]);
</script>

<style scoped></style>
