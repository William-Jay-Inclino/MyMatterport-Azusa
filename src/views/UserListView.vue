<template>
  <div class="columns pattern-background">
    <div class="column is-2 menu-width">
      <TheLeftSidebar />
    </div>
    <UserList @add-user="addUser" @delete-user="deleteUser" />
  </div>
</template>

<script setup lang="ts">
import UserList from "@/components/user/UserList.vue";
import UserAdminModal from "@/components/user/UserAdminModal.vue";
import { useUserStore } from "@/stores/user";
import { useProgrammatic } from "@oruga-ui/oruga-next";
import { computed } from "vue";
import { User } from "@/stores/types";
import TheLeftSidebar from "@/components/TheLeftSidebar.vue";

const { oruga } = useProgrammatic();
const userStore = useUserStore();

// Opens modal to add Admin
const addUser = (payload: { action: string; User }) => {
  userStore.showUserForm = true;
  oruga.modal.open({
    component: UserAdminModal,
    trapFocus: true,
  }),
    console.log("Adds User", payload.User);
};

// Gets number of Checked User
const checkedUserLength = computed(() => {
  return userStore.checkedUsers.length;
});

// Deletes checked Users
const deleteUser = (payload: { action: string; user: User }) => {
  for (let i = 0; i < checkedUserLength.value; i++) {
    userStore.selectedUser = payload.user[i];
    userStore.deleteUser(userStore.selectedUser);
    userStore.selectedUser = {};
  }
  console.log("Successfully Deleted Checked Users", payload.user);
};
</script>

<style scoped>
.pattern-background {
  margin-top: 0px;
  max-width: 100%;
  height: 100vh;
}
</style>
