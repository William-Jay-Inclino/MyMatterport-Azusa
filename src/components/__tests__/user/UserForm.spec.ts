import { describe, it, expect, beforeEach, vi } from "vitest" 
import type WrapperLike from "@vue/test-utils/dist/interfaces/wrapperLike"
import { createPinia, setActivePinia } from "pinia"
import { shallowMount } from "@vue/test-utils"
import { createTestingPinia } from '@pinia/testing'

import UserListView from "@/views/UserListView.vue"
import UserForm from "@/components/user/UserForm.vue"
import type { User } from "@/stores/types";
import { useUserStore } from "@/stores/user";
import { faker } from "@faker-js/faker";
 
describe("=== Components ===", () => {
    let wrapper: WrapperLike
    
    setActivePinia(createPinia())
    wrapper = shallowMount(UserForm, {
        global: {
            plugins: [createTestingPinia()],
        }
    })

    it("Should Mount UserForm.vue", () => {
        expect(wrapper).toBeTruthy()
    })

    it("Should saveClick() mode = add", async() => {
        const userStore = useUserStore()
        userStore.mode = 'add'
        const addUserSpy = vi.spyOn(userStore, 'addUser') //spy addUser() function in store
        //@ts-ignore
        await wrapper.vm.saveClick() //call saveClick() function in component 
        expect(addUserSpy).toHaveBeenCalledOnce() 
    })

    it("Should saveClick() mode = edit", async() => {
        const userStore = useUserStore()
        userStore.mode = 'edit'
        const updateUserSpy = vi.spyOn(userStore, 'updateUser')
        //@ts-ignore
        await wrapper.vm.saveClick()
        expect(updateUserSpy).toHaveBeenCalledOnce()
    })

    it("Should deleteClick()", async() => {
        const userStore = useUserStore()
        const deleteUserSpy = vi.spyOn(userStore, 'deleteUser')
        //@ts-ignore
        await wrapper.vm.deleteClick()
        expect(deleteUserSpy).toHaveBeenCalledOnce()
    })

    it("Should cancelClick()", async() => {
        const userStore = useUserStore()
        const resetFormSpy = vi.spyOn(userStore, 'resetForm')
        //@ts-ignore
        await wrapper.vm.cancelClick()
        expect(resetFormSpy).toHaveBeenCalledOnce()
    })
    
})


// it("Should add user", async() => {
//     const user_store = useUserStore() 
//     user_store.mode = 'add'
//     const total_users_before = user_store.users.length

//     // faker user
//     const user = {} as User 
//     user.uuid = faker.datatype.uuid()
//     user.first_name = faker.name.firstName()
//     user.last_name = faker.name.lastName()
//     user.email = faker.internet.email()
//     user.mobile = faker.phone.number()
//     user.city = faker.address.cityName()
//     user_store.userForm = user
//     // end

//     // get element wrapper
//     const first_name_wrapper = wrapper.get(`[data-test="input-first-name"]`)
//     const last_name_wrapper = wrapper.get(`[data-test="input-last-name"]`)
//     const email_wrapper = wrapper.get(`[data-test="input-email"]`)
//     const mobile_wrapper = wrapper.get(`[data-test="input-mobile"]`)
//     // end

//     // populate input fields
//     await first_name_wrapper.trigger('input')
//     await last_name_wrapper.trigger('input')
//     await email_wrapper.trigger('input')
//     await mobile_wrapper.trigger('input')
//     // end

//     const btn_save = wrapper.get(`[data-test="button-save"]`)
//     await btn_save.trigger("click")

//     expect(user_store.users.length).toBe(total_users_before + 1) // expect current no. of users to be greater than no. of users previously

//     // expect user added to be in users array
//     console.log('users', user_store.users)
//     const user_added = user_store.users.find( (u: User) => (u.first_name == user.first_name && u.last_name == user.last_name && u.email == user.email && u.mobile == user.mobile)) 
//     console.log('user_added', user_added)
//     expect(user_added).not.toBe(null)
//     // end
// })