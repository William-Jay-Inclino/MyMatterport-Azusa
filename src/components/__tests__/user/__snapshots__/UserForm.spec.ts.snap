// Vitest Snapshot v1

exports[`=== UI Elements === > Should find button-cancel 1`] = `
DOMWrapper {
  "isDisabled": [Function],
  "wrapperElement": <o-button
    data-test="button-cancel"
  >
    Cancel
  </o-button>,
}
`;

exports[`=== UI Elements === > Should find button-delete 1`] = `
DOMWrapper {
  "isDisabled": [Function],
  "wrapperElement": <o-button
    data-test="button-delete"
  >
    Delete
  </o-button>,
}
`;

exports[`=== UI Elements === > Should find button-save 1`] = `
DOMWrapper {
  "isDisabled": [Function],
  "wrapperElement": <o-button
    data-test="button-save"
  >
    Save
  </o-button>,
}
`;

exports[`=== UI Elements === > Should find input-email 1`] = `
DOMWrapper {
  "isDisabled": [Function],
  "wrapperElement": <o-input
    data-test="input-email"
  />,
}
`;

exports[`=== UI Elements === > Should find input-first-name 1`] = `
DOMWrapper {
  "isDisabled": [Function],
  "wrapperElement": <o-input
    data-test="input-first-name"
  />,
}
`;

exports[`=== UI Elements === > Should find input-last-name 1`] = `
DOMWrapper {
  "isDisabled": [Function],
  "wrapperElement": <o-input
    data-test="input-last-name"
  />,
}
`;

exports[`=== UI Elements === > Should find input-mobile 1`] = `
DOMWrapper {
  "isDisabled": [Function],
  "wrapperElement": <o-input
    data-test="input-mobile"
  />,
}
`;
